src/runtime/virtcontainers/container.go

```diff
@@ -869,6 +869,23 @@ func (c *Container) create(ctx context.Context) (err error) {
 		}
 	}
 
+	// If cold-plug we've attached the devices already, do not try to
+	// attach them a second time.
+	coldPlugVFIO := (c.sandbox.config.HypervisorConfig.ColdPlugVFIO != config.NoPort)
+	if coldPlugVFIO {
+		var cntDevices []ContainerDevice
+		for _, dev := range c.devices {
+			if strings.HasPrefix(dev.ContainerPath, vfioPath) {
```

Comment:
Does the cold-plug implementation handle **ALL** vfio devices, including network, gpu, nvme etc? I'm asking because the change here assumes that all vfio devices are already code-plugged.

Reply:
That's a good point. At sandbox creation time I am "marking" cold-plugged devices I need to add an additional check if they are VFIO && cold-plugged.

Reply:
Ok thinking more about this, if we set `cold_plug_vfio="root-port"` ALL vfio devices are cold-plugged to the `root-port` and hence all VFIO devices need to be removed in the container create stage. 

This PR https://github.com/kata-containers/kata-containers/pull/4699/files#diff-e6df7a8b934e487feec13185e69c9bcb0e256e513bb792141ec9162d5edad399 will introduce selectively choose were to attach each VFIO device.

Reply:
We currently have no way of telling how individual devices are going to be hot|cold plugged. Added CDI in this PR to exactly solve this with the aforementioned PR.